SymPEP-XXXX.md

```diff
@@ -0,0 +1,136 @@
+
+
+```python
+from sympy import Expr, Symbol
+
+def differentiate(expr, var):
+    """
+    Differentiate a SymPy expression with respect to a variable.
+    """
+    if isinstance(expr, Expr) and isinstance(var, Symbol):
+        return expr.diff(var)
+    else:
+        raise TypeError("expr must be a SymPy expression and var must be a SymPy symbol")
+
+def simplify_expressions(expressions):
+    """
+    Simplify a list of SymPy expressions.
+    """
+    if isinstance(expressions, list):
+        if all(isinstance(expr, Expr) for expr in expressions):
+            return [expr.simplify() for expr in expressions]
+    else:
+        raise TypeError("expressions must be a list of SymPy expressions")
+```
+
+which is more verbose and less efficient than the static typed versions, because of added runtime type checks.
+
+## Backwards Compatibility
+
+Introducing static typing to SymPy may break backwards compatibility for code that relies on dynamically typed behavior. Users who heavily depend on runtime type inference might need to update their code to match the new type annotations.
+
+## Detailed Description
+
+This SymPEP proposes the gradual introduction of static type annotations using tools like Python's `typing` module and third-party type checkers such as `mypy` or `pyright`. The process will involve identifying critical modules, functions, and classes to begin the static typing integration. We will define guidelines for annotating function signatures, class attributes, and return types. The goal is to maintain compatibility with existing dynamically typed code while allowing for a smooth transition.
+
+Authors of new classes, functions, or modules in SymPy should be encouraged to write their code with static typing, unless they encounter a situation where achieving this is difficult or not possible. In such cases, they are expected to include a comment explaining the reasons preventing static typing.
+
+Introducing new dynamically typed code initially can lead to the accumulation of technical debt. It's recognized that migrating existing dynamically typed code later is significantly more challenging than initially writing code with static typing. Therefore, introducing new dynamically typed code should be approached cautiously.
+
+It's important to note that tools like mypy and pyright are capable of inferring types, facilitating the incorporation of static typing without a steep learning curve. By adding a few annotations, code can be made cleaner and the transition to static typing across the entire codebase can be eased.
+
+For functions, classes, and modules internally used by SymPy that currently feature unnecessary dynamic type checks, a shift towards static typing should be promoted. This transition can help eliminate these unnecessary checks and subsequently enhance the overall performance of the SymPy codebase.
```

Comment:
Who deems these checks unnecessary? If we want a function/method to be duck-typed, I believe we should be allowed to do so. Duck-typing has been the design paradigm preached by the Python community sense I can remember and it is a feature, not a bug.

Reply:
I agree that we should not drop dynamic type checks because of adding static typing. When not using an editor supporting this (you have no idea how many students edit their code in whatever the system offers), there should still be sensible error messages, not the code crashing later with a seemingly unrelated error message. 

If nothing else, this will probably reduce the number of issues opened that boils down to the user doing something wrong.

Reply:
Note also that if the type hints are used consistently on user facing functions then an editor like vscode will already warn the user about passing the wrong type into a function before the code even runs.

Reply:
Python allows to inspect type annotations in runtime

```python
def square(x: int):
    return x * x

square.__annotations__
# {'x': int}
```

such that we can implement something like `@validate` such that

```
# Function for users that raises on isinstance(x, int)
@validate
def square(x: int):
    return x * x

# Function for internal use
def square(x: int):
    return x * x
```

So even if we want to do runtime validation, 
It should be better done like above, than repeating `isinstance` checks manually,
which often makes the code more complicated with validation code.

Reply:
I totally agree. Just wanted to point out that not all people use editors with type hinting (and/or do not understand how to benefit from them).

Reply:
> not all people use editors with type hinting

Agreed. I make very heavy use of ipython for example which does not support this.

Note though that the sentence in the PEP that we are all commenting on here explicitly says (emphasis added):

> For functions, classes, and modules **INTERNALLY** used by SymPy that currently feature unnecessary dynamic type checks, a shift towards static typing should be promoted.

The example here is a good one:
https://github.com/sympy/sympy/blob/b0dcb5af49e7289680d9789d292197675e40490d/sympy/polys/polyclasses.py#L167-L171
In [gh-25651](https://github.com/sympy/sympy/pull/25651/files) I found bugs by enabling that check. The check is too expensive to be used at runtime though so it was commented out before merge. Almost everything that it checks is something that could be verified by a static type checker but doing that check at runtime means recursively calling `isinstance` down through a potentially large data structure. The class in question is purely internal and not something that any ordinary SymPy user would ever see directly.

Note also that just attempting to verify that the codebase does indeed only use those types in this particular function is something that consumed development time. The whole PR [gh-25651](https://github.com/sympy/sympy/pull/25651/files) would be completely unnecessary if the code in question just used type hints but as it is if I want to verify what the types are than I have to add runtime checks and ensure that the entire CI test suite completes successfully (and then fix the bugs that show up).